src/pgen/cluster/stellar_feedback.cpp

```diff
@@ -0,0 +1,155 @@
+  using parthenon::IndexDomain;
+  using parthenon::IndexRange;
+  using parthenon::Real;
+
+  auto hydro_pkg = md->GetBlockData(0)->GetBlockPointer()->packages.Get("Hydro");
+
+  if (disabled_) {
+    // No stellar feedback, return
+    return;
+  }
+
+  // Grab some necessary variables
+  const auto &prim_pack = md->PackVariables(std::vector<std::string>{"prim"});
+  const auto &cons_pack = md->PackVariables(std::vector<std::string>{"cons"});
+  IndexRange ib = md->GetBlockData(0)->GetBoundsI(IndexDomain::interior);
+  IndexRange jb = md->GetBlockData(0)->GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = md->GetBlockData(0)->GetBoundsK(IndexDomain::interior);
+  const auto nhydro = hydro_pkg->Param<int>("nhydro");
+  const auto nscalars = hydro_pkg->Param<int>("nscalars");
+
+  // const auto gm1 = (hydro_pkg->Param<Real>("AdiabaticIndex") - 1.0);
+  const auto units = hydro_pkg->Param<Units>("units");
+  const auto mbar = hydro_pkg->Param<Real>("mu") * units.mh();
+  const auto mbar_over_kb = hydro_pkg->Param<Real>("mbar_over_kb");
+
+  const auto mass_to_energy = efficiency_ * SQR(units.speed_of_light());
+  const auto stellar_radius = stellar_radius_;
+  const auto temperature_threshold = temperatue_threshold_;
+  const auto number_density_threshold = number_density_threshold_;
+
+  const auto eos = eos_;
+
+  ////////////////////////////////////////////////////////////////////////////////
+
+  // Constant volumetric heating
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "StellarFeedback::FeedbackSrcTerm", parthenon::DevExecSpace(),
+      0, cons_pack.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int &b, const int &k, const int &j, const int &i) {
+        auto &cons = cons_pack(b);
+        auto &prim = prim_pack(b);
+        const auto &coords = cons_pack.GetCoords(b);
+
+        const auto x = coords.Xc<1>(i);
+        const auto y = coords.Xc<2>(j);
+        const auto z = coords.Xc<3>(k);
+
+        const auto r = sqrt(x * x + y * y + z * z);
+        if (r > stellar_radius) {
+          return;
+        }
+
+        auto number_density = prim(IDN, k, j, i) / mbar;
+        if (number_density < number_density_threshold) {
+          return;
+        }
+
+        auto temp = mbar_over_kb * prim(IPR, k, j, i) / prim(IDN, k, j, i);
+        if (temp > temperature_threshold) {
+          return;
+        }
+
+        // All conditions to convert mass to energy are met
+        const auto cell_delta_rho = number_density_threshold * mbar - prim(IDN, k, j, i);
+
+        // First remove density at fixed temperature
+        AddDensityToConsAtFixedVelTemp(cell_delta_rho, cons, prim, eos.GetGamma(), k, j,
+                                       i);
+        //  Then add thermal energy
+        const auto cell_delta_mass = -cell_delta_rho * coords.CellVolume(k, j, i);
+        const auto cell_delta_energy_density =
+            mass_to_energy * cell_delta_mass / coords.CellVolume(k, j, i);
```

Comment:
We're multiplying by cell volume and then dividing by cell volume? Can't we do this:
```suggestion
        const auto cell_delta_energy_density =
            -mass_to_energy * cell_delta_rho
```

Reply:
Yes, I fixed this. In my first implementation everything was very verbose.